#**Session Flow**



##**👨🏻‍🎓Learning Objective**
* Introduction
* Theme
* Primary Goals

##**Plotly: Interactive Data Visualization Library**
* Introduction to plotly
* Load data from google sheets
* Plotting graphs hist,pie,bar,scatter,line,heatmap,guage chart , tree chart
* How to create subplots?
* Interactivity in Plotly
* Activity: True or False



##**📝 Summary**

* What did we learn?
* Shortcomings & Challenges of using Plotly
* Best Practices & Tips for effective visualizations

##**🔗 Enhance your Knowledge**

* Mnemonic
* Additional Reading
* Additional Resources

## **Try it Yourself**

* Activities
* Social Engagement



# **👨🏻‍🎓 Learning Objective 👨🏻‍🎓**

**Introduction**

👋 Welcome to the lesson on Data Visualization Tips and Best Practices! 📊💡

This lesson will explore useful tips and best practices to create effective and engaging data visualizations. Here are some emojis to help illustrate these tips:

1️⃣ 🎯 Determine the purpose and audience of your visualization. It's important to know what message you want to convey and who will be viewing your visualization.

2️⃣ 🖼️ Choose the appropriate visualization type. Different data sets require different visualization types, such as bar charts 📊, line graphs 📈, and scatter plots 🌟.

3️⃣ 🔍 Simplify and declutter your visualization. Keep your visualizations simple and easy to understand by focusing on the key data points and avoiding unnecessary elements.

4️⃣ 🎨 Use appropriate colors and fonts. Colors and fonts can enhance the visual appeal of your visualization and help convey information.

5️⃣ 🏷️ Label and title your visualization appropriately. Label all axes, legends, and data points, and use descriptive titles that accurately summarize the information.

6️⃣ 📝 Include context and annotations. Provide context to your data by including annotations or additional information to help viewers understand the significance of the data.

7️⃣ 🔄 Test and iterate your visualization. Test your visualization with different audiences and make changes as necessary until you achieve the desired impact.
By following these tips and best practices, you'll be able to create wonderful and effective data visualizations that communicate your message clearly and efficiently. 🤩 So, let's get started!


##**Theme**
##**📊💡 Theme: "Visualize to Empower: Unleash Data's Impact!" 📊**

🌟 In today's data-driven world, understanding data visualization is key to unleashing the true potential of data. 🌟 By learning to create compelling visualizations, you can unlock insights, enhance communication, engage audiences, empower decision-making, promote data literacy, and drive innovation. 🚀💪 Embrace the art of data visualization to communicate data's stories with impact and make a lasting impression in the realm of data-driven communication! 🤩🎨

**Primary Goals**

1️⃣ 🥅👀📊
Goal 1: Choose the right visualization for your data. Our aim is to explore different types of visualizations and identify when they are most appropriate to use based on your data and message. By the end of this section, you will be able to select the most effective visualization type to communicate your message.

2️⃣ 🥅🛑📈
Goal 2: Identify common data visualization challenges and their solutions. We'll cover obstacles such as cluttered visualizations, lack of context, and other pitfalls to avoid. By the end of this section, you will be able to avoid common mistakes and create more effective visualizations.

3️⃣ 🥅🎨📝🤔
Goal 3: Apply best practices for data visualization design. We'll discuss tips on simplifying and decluttering your visualizations, using appropriate colors and fonts, labeling and titling appropriately, including context and annotations, and testing and iterating. By the end of this section, you'll be able to create visually appealing and effective data visualizations.
By achieving these goals, you'll be able to create impactful data visualizations that effectively communicate your message. So, let's get started and learn how to make visually engaging and effective data visualizations! 🤓


# Plotly: Interactive Data Visualization Library

## **What is Plotly?**
Plotly is a powerful Python library for creating interactive and aesthetically pleasing visualizations. It allows users to generate static, animated, and interactive plots for web-based or standalone applications.

Plotly supports:
- Integration with Python, R, JavaScript, MATLAB, and Julia.
- A wide range of visualization types, from simple scatter plots to complex 3D plots and dashboards.
- Interactive features like zooming, panning, and hover annotations.

---

## **Key Features of Plotly**
1. **Interactivity**:
   - Supports zoom, pan, and hover for data exploration.
   - Allows for drill-downs and detailed views.

2. **Wide Range of Visualizations**:
   - Basic charts: Line, Bar, Scatter, Pie.
   - Advanced charts: Tree Maps, Sunbursts, 3D Scatter Plots.
   - Statistical charts: Histograms, Box Plots, Distplots.
   - Business charts: Waterfall, Funnel, and Gauge Charts.
   - Heatmaps, Bubble Charts, and more.

3. **Customization**:
   - Highly customizable layouts, colors, and markers.
   - Responsive designs for various screen sizes.

4. **Integration**:
   - Seamlessly integrates with Jupyter Notebooks and Dash for web-based applications.
   - Works well with pandas and NumPy for data handling.

5. **Web-based Sharing**:
   - Visualizations can be saved as HTML files or embedded into web applications.
   - Easy sharing through static or interactive formats.

6. **Advanced Analytics**:
   - Built-in support for statistical and financial charts.
   - Capable of creating 3D plots for in-depth visualizations.

---

## **Why Use Plotly?**
1. **User-Friendly Interactivity**:
   - Unlike static libraries like Matplotlib or Seaborn, Plotly provides an interactive experience, enabling deeper insights through direct user engagement.

2. **Dynamic Dashboards**:
   - Perfect for real-time monitoring or presentation of Key Performance Indicators (KPIs) in dashboards.

3. **Customization & Professional Aesthetics**:
   - Enhances the visual appeal and provides better control over chart appearance compared to traditional libraries.

4. **Web-Ready Visualizations**:
   - Easily share plots online or embed them in websites and applications.

5. **Broad Application Range**:
   - Suitable for exploratory data analysis, business intelligence, reporting, and educational purposes.

---

## **What Can Be Done with Plotly?**
1. **Basic Charts**:
   - Line, Bar, Scatter, Pie, and Histogram.

2. **Statistical Charts**:
   - Box Plots, Violin Plots, Distplots.

3. **3D Visualizations**:
   - 3D Scatter, 3D Surface, and 3D Mesh Plots.

4. **Business Analytics**:
   - Waterfall, Funnel, and Gauge Charts.

5. **Hierarchical Charts**:
   - Tree Maps, Sunburst Charts, and Packed Circle Charts.

6. **Heatmaps**:
   - Visualize correlations or intensities.

7. **Custom Dashboards**:
   - Create interactive dashboards with `Dash`, a Python framework built on Plotly.

---

## **Advantages of Plotly**
- Interactive and dynamic features for a superior user experience.
- Advanced visualization options unavailable in libraries like Matplotlib or Seaborn.
- Ideal for real-time data dashboards and presentations.
- Great for web applications with easy HTML export and embedding.

---

Plotly is an essential tool for anyone looking to enhance their data visualization capabilities, whether for business insights, academic research, or professional presentations.


**Credential Link for JSON File:** https://drive.google.com/file/d/1rJIc6U6rx9z6IuQS1QAWA9xfpyIXJ3pi/view?usp=sharing

**Google Sheet Link:** https://docs.google.com/spreadsheets/d/1s54ztt0D7dvpgmsdIgsYkoEGwHCA6Ndn0G2VyNhW19E/edit?usp=sharing

**Google Service Account ID:** python-api@python-api-project-443806.iam.gserviceaccount.com

# How to connect to GCP

1. First of all we need to create a project on GCP Console
2. Second, you need to a haver service account which should have the api permission
3. we will enable the api services for google drive and google spreadsheets
4. download the json key from the keys section of the credentials in the service account

# Load data from google sheet

In [1]:
import gspread
from google.oauth2.service_account import Credentials
import pandas as pd

# Authenticate and initialize client
scopes = ["https://www.googleapis.com/auth/spreadsheets"]
creds = Credentials.from_service_account_file(
    "/content/ace-memento-465318-u0-ea7242b2370f.json",
    scopes=scopes
)
client = gspread.authorize(creds)

# Open the Google Sheet by its ID
sheet_id = "1s54ztt0D7dvpgmsdIgsYkoEGwHCA6Ndn0G2VyNhW19E" # Take id from your google sheets
sheet = client.open_by_key(sheet_id)

# Select the first worksheet
worksheet = sheet.get_worksheet(0)
# print(sheet)


# Import data from google sheet as csv

In [2]:
from gspread_dataframe import get_as_dataframe

# Convert the worksheet data to a DataFrame
data = get_as_dataframe(worksheet)
data.info()
# data.dropna(how='all')
# data.head()

<class 'pandas.core.frame.DataFrame'>
Index: 554 entries, 0 to 553
Data columns (total 28 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Unnamed: 0                               554 non-null    float64
 1   Country                                  554 non-null    object 
 2   Gross tertiary education enrollment (%)  554 non-null    float64
 3   Population                               554 non-null    float64
 4   Title                                    554 non-null    object 
 5   Unemployment rate                        554 non-null    float64
 6   Urban_population                         554 non-null    float64
 7   Youtuber                                 554 non-null    object 
 8   category                                 554 non-null    object 
 9   channel_type                             554 non-null    object 
 10  channel_type_rank                        554 non-null  

In [3]:
# from gspread_dataframe import get_as_dataframe

# Convert the worksheet data to a DataFrame
# data = get_as_dataframe(worksheet)
data = data.dropna(how='all').dropna(axis=1, how='all')
data


,Unnamed: 0,Country,Gross tertiary education enrollment (%),Population,Title,Unemployment rate,Urban_population,Youtuber,category,channel_type,...,lowest_yearly_earnings,rank,subscribers,subscribers_for_last_30_days,uploads,video views,video_views_for_the_last_30_days,video_views_rank,engagement_rate,views_per_upload
0,0.0,India,28.1,1.366418e+09,T-Series,5.36,471031528.0,T-Series,Music,Music,...,6800000.0,1.0,245000000.0,2000000.0,20082.0,2.280000e+11,2.258000e+09,1.0,0.816327,1.135345e+07
1,2.0,United States,88.2,3.282395e+08,MrBeast,14.70,270663028.0,MrBeast,Entertainment,Entertainment,...,4000000.0,3.0,166000000.0,8000000.0,741.0,2.836884e+10,1.348000e+09,48.0,4.819277,3.828454e+07
2,3.0,United States,88.2,3.282395e+08,Cocomelon - Nursery Rhymes,14.70,270663028.0,Cocomelon - Nursery Rhymes,Education,Education,...,5900000.0,4.0,162000000.0,1000000.0,966.0,1.640000e+11,1.975000e+09,2.0,0.617284,1.697723e+08
3,4.0,India,28.1,1.366418e+09,SET India,5.36,471031528.0,SET India,Shows,Entertainment,...,5500000.0,5.0,159000000.0,1000000.0,116536.0,1.480000e+11,1.824000e+09,3.0,0.628931,1.269994e+06
4,8.0,Russia,81.9,1.443735e+08,Like Nastya Vlog,4.59,107683889.0,Like Nastya,People & Blogs,People,...,146800.0,9.0,106000000.0,100000.0,493.0,9.047906e+10,4.894700e+07,630.0,0.094340,1.835275e+08
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
549,989.0,United States,88.2,3.282395e+08,Migos ATL,14.70,270663028.0,Migos ATL,Music,Entertainment,...,148200.0,990.0,12400000.0,100000.0,99.0,6.993406e+09,4.941200e+07,833.0,0.806452,7.064047e+07
550,990.0,Brazil,51.3,2.125594e+08,Natan por AÔø,12.08,183241641.0,Natan por AÔø,Sports,Entertainment,...,1700000.0,991.0,12300000.0,700000.0,1200.0,9.029610e+09,5.525130e+08,525.0,5.691057,7.524675e+06
551,991.0,India,28.1,1.366418e+09,Free Fire India Official,5.36,471031528.0,Free Fire India Official,People & Blogs,Games,...,194200.0,992.0,12300000.0,300000.0,1500.0,1.674410e+09,6.473500e+07,6141.0,2.439024,1.116273e+06
552,993.0,Sweden,67.0,1.028545e+07,RobTopGames,6.48,9021165.0,RobTopGames,Gaming,Games,...,11600.0,994.0,12300000.0,100000.0,39.0,3.741235e+08,3.871000e+06,35112.0,0.813008,9.592910e+06


In [4]:
data.drop('Unnamed: 0',axis=1,inplace=True)

In [5]:
data

,Country,Gross tertiary education enrollment (%),Population,Title,Unemployment rate,Urban_population,Youtuber,category,channel_type,channel_type_rank,...,lowest_yearly_earnings,rank,subscribers,subscribers_for_last_30_days,uploads,video views,video_views_for_the_last_30_days,video_views_rank,engagement_rate,views_per_upload
0,India,28.1,1.366418e+09,T-Series,5.36,471031528.0,T-Series,Music,Music,1.0,...,6800000.0,1.0,245000000.0,2000000.0,20082.0,2.280000e+11,2.258000e+09,1.0,0.816327,1.135345e+07
1,United States,88.2,3.282395e+08,MrBeast,14.70,270663028.0,MrBeast,Entertainment,Entertainment,1.0,...,4000000.0,3.0,166000000.0,8000000.0,741.0,2.836884e+10,1.348000e+09,48.0,4.819277,3.828454e+07
2,United States,88.2,3.282395e+08,Cocomelon - Nursery Rhymes,14.70,270663028.0,Cocomelon - Nursery Rhymes,Education,Education,1.0,...,5900000.0,4.0,162000000.0,1000000.0,966.0,1.640000e+11,1.975000e+09,2.0,0.617284,1.697723e+08
3,India,28.1,1.366418e+09,SET India,5.36,471031528.0,SET India,Shows,Entertainment,2.0,...,5500000.0,5.0,159000000.0,1000000.0,116536.0,1.480000e+11,1.824000e+09,3.0,0.628931,1.269994e+06
4,Russia,81.9,1.443735e+08,Like Nastya Vlog,4.59,107683889.0,Like Nastya,People & Blogs,People,25.0,...,146800.0,9.0,106000000.0,100000.0,493.0,9.047906e+10,4.894700e+07,630.0,0.094340,1.835275e+08
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
549,United States,88.2,3.282395e+08,Migos ATL,14.70,270663028.0,Migos ATL,Music,Entertainment,171.0,...,148200.0,990.0,12400000.0,100000.0,99.0,6.993406e+09,4.941200e+07,833.0,0.806452,7.064047e+07
550,Brazil,51.3,2.125594e+08,Natan por AÔø,12.08,183241641.0,Natan por AÔø,Sports,Entertainment,172.0,...,1700000.0,991.0,12300000.0,700000.0,1200.0,9.029610e+09,5.525130e+08,525.0,5.691057,7.524675e+06
551,India,28.1,1.366418e+09,Free Fire India Official,5.36,471031528.0,Free Fire India Official,People & Blogs,Games,69.0,...,194200.0,992.0,12300000.0,300000.0,1500.0,1.674410e+09,6.473500e+07,6141.0,2.439024,1.116273e+06
552,Sweden,67.0,1.028545e+07,RobTopGames,6.48,9021165.0,RobTopGames,Gaming,Games,69.0,...,11600.0,994.0,12300000.0,100000.0,39.0,3.741235e+08,3.871000e+06,35112.0,0.813008,9.592910e+06


# 📈 Histogram

## **What is Histogram?**
A histogram shows the frequency distribution of a numeric variable by dividing the data into bins.

## **Why Histogram**
- Visualize the distribution of numerical data.
- Identify patterns like skewness or modality.
- Detect outliers in the dataset.

## **Syntax**
```python
import plotly.express as px

fig = px.histogram(data_frame, x="numeric_column", nbins=20, title="Histogram")
fig.show()


In [7]:
import numpy as np
import pandas as pd
import plotly.express as px

**What is the distribution of the engagement_rate in the dataset?**

In [ ]:
fig = px.histogram(data, x='engagement_rate',nbins=80,title="Histogram of Engagement Rates",color_discrete_sequence=['green'])

# Customize size and font
fig.update_layout(height=500,  # Set the height of the figure
    width=700,  # Set the width of the figure
    title_font=dict(size=20, color='darkblue'),  # Title font settings
    font=dict(size=14,color='darkblue'),  # General font size for axes and labels
    bargap=0.01)  # Gap between bars

fig.show()

 **What is the variability in subscribers among channels?**

In [ ]:
fig_subscribers_histogram = px.histogram(
    data, x='subscribers', nbins=300,
    title="Histogram of Subscriber Count"
)

# Customize size and font
fig_subscribers_histogram.update_layout( height=600,  # Set the height of the figure
    width=900,  # Set the width of the figure
    title_font=dict(size=20, color='darkblue'),  # Title font settings
    font=dict(size=14,color='darkblue'),  # General font size for axes and labels
    bargap=0.01  # Gap between bars
)

fig_subscribers_histogram.show()


# Insights
* **Skewed Distribution:** The majority of subscriber counts are concentrated below 50 million, indicating a skewed distribution.
* **Decline in Subscriber Counts:** A sharp decline is observed as subscriber counts increase, with very few entities exceeding 100 million.
* **Presence of Outliers:** Outliers or exceptional performers are present in the higher subscriber ranges.
* **Growth Opportunities:** The data suggests significant market saturation at lower levels and opportunities for growth in the higher ranges.

# 📊 Count Plot (Bar Chart)

## **What is countplot?**
A bar chart represents categorical data using rectangular bars where the length corresponds to the frequency or value.

## **Why is countplot?**
- Compare frequencies of different categories.
- Analyze trends across categorical data.
- Useful for visualizing survey responses or group comparisons.

## **Syntax**
```python
import plotly.express as px

fig = px.bar(data_frame, x="category_column", title="📊 Count Plot (Bar Chart)")
fig.show()


**How many YouTube channels belong to each category?**

In [ ]:
data['category'].value_counts().reset_index()

,category,count
0,Entertainment,145
1,Music,110
2,People & Blogs,72
3,Gaming,49
4,Comedy,40
5,Education,33
6,Film & Animation,28
7,News & Politics,21
8,Howto & Style,13
9,Shows,12


In [ ]:
df=data['category'].value_counts().reset_index()
fig_category_count = px.bar(
    df,
    y='count', x='category',
    title="Count Plot: Number of YouTube Channels per Category",
    labels={'index': 'Category', 'category': 'Count'}
)

fig_category_count .update_layout( height=600,  # Set the height of the figure
    width=900,  # Set the width of the figure
    title_font=dict(size=20, color='darkblue'),  # Title font settings
    font=dict(size=14,color='darkblue'),  # General font size for axes and labels
    bargap=0.15  # Gap between bars
)
# Apply rainbow colors

fig_category_count.show()


### **Insights**
* **Dominant Categories:** Entertainment and Music categories have the highest number of YouTube channels, indicating their widespread popularity and appeal.
* **Moderate Representation:** Categories like People & Blogs, Gaming, and Comedy also have a significant presence, reflecting diverse content consumption preferences.
* **Niche Categories:** Topics such as Autos & Vehicles, Nonprofits & Activism, and Movies have relatively fewer channels, suggesting specialized or limited audiences.
* **Growth Potential:** Categories with lower representation, such as Science & Technology and Sports, could offer opportunities for creators to tap into underserved niches.







**How many YouTube channels are there for each channel type?**

In [ ]:
data.columns

Index(['Country', 'Gross tertiary education enrollment (%)', 'Population',
       'Title', 'Unemployment rate', 'Urban_population', 'Youtuber',
       'category', 'channel_type', 'channel_type_rank', 'country_rank',
       'created_date', 'created_month', 'created_year',
       'highest_monthly_earnings', 'highest_yearly_earnings',
       'lowest_monthly_earnings', 'lowest_yearly_earnings', 'rank',
       'subscribers', 'subscribers_for_last_30_days', 'uploads', 'video views',
       'video_views_for_the_last_30_days', 'video_views_rank',
       'engagement_rate', 'views_per_upload'],
      dtype='object')

In [ ]:
# how to import the different data like csv, .xlsx
# pandas how to import csv
data = pd.read_csv("")
# pandas how to import xlsx
data = pd.read_excel("")

In [ ]:
df=data['channel_type'].value_counts().reset_index()
fig_channel_type_count = px.bar(
    df,
    x='count', y='channel_type',
    title="Count Plot: Distribution of Channel Types")
fig_channel_type_count.show()


### **Insights**
* **Entertainment Dominates:** Entertainment channels have the highest representation, indicating their massive appeal and popularity among creators and audiences.
* **Music as a Strong Contender:** Music channels follow closely, showing their global demand and wide audience reach.
* **Moderate Growth Categories:** Games, People, and Education have a considerable number of channels, reflecting their steady growth and interest.
* **Underrepresented Niches:** Categories like Autos, Nonprofit, and Animals are less represented, suggesting potential opportunities for creators in these areas.

# 📦 Box Plot

## **What**
A box plot shows the distribution of a dataset, highlighting medians, quartiles, and potential outliers.

## **Why**
- Summarize the distribution of numerical data.
- Compare distributions across groups.
- Detect outliers in the data.

## **Syntax**
```python
import plotly.express as px

fig = px.box(data_frame, x="category_column", y="numeric_column", title="📦 Box Plot")
fig.show()


**What is the distribution of engagement rates for YouTube channels in each country?**

In [8]:
fig_engagement_country = px.box(
    data, x='Country', y='engagement_rate',
    title="Box Plot: Engagement Rate by Country",
    color='Country'
)
fig_engagement_country.show()

**Insights for Engagement Rate by Country**
* **High Engagement:** Countries like India and a few others show significantly high engagement rates, with some outliers exceeding 30%.
* **Global Variability:** There is substantial variation in engagement rates across countries, indicating diverse audience interactions.
* **Clusters:** Most countries have engagement rates concentrated in the lower range (below 10%), with few exceptions.
* **Outliers:** Some countries exhibit extreme engagement rates, highlighting unique user behavior or market-specific trends.

**How do video views vary across different YouTube categories?**

In [ ]:
fig_video_views_category = px.box(
    data, x='category', y='video views',
    title="Box Plot: Video Views by Category",
    color='category'
)
fig_video_views_category.show()


**Insights for Video Views by Category**
* **Top Categories:** Music and Entertainment categories dominate in terms of video views, with some reaching over 200 billion views.
* **Steady Performance:** Categories like Education and People & Blogs exhibit steady views but are outpaced by top categories.
* **Niche Content:** Categories like Science & Technology, Autos & Vehicles, and Nonprofits & Activism have lower video views, suggesting smaller audiences.
* **Outliers:** Certain categories have extreme outliers, indicating viral or exceptionally popular content.

# 📈 Line Chart

## **What**
A line chart is used to visualize trends over time or continuous data.

## **Why**
- Analyze changes over time.
- Identify trends, patterns, or anomalies.
- Useful for time-series data like sales, stock prices, or weather.

## **Syntax**
```python
import plotly.express as px

fig = px.line(data_frame, x="time_column", y="numeric_column", title="📈 Line Chart")
fig.show()


**What is the yearly trend in the number of YouTube channels created?**

In [ ]:
data['created_year'].value_counts().reset_index().sort_values(by='created_year',ignore_index=True)

,created_year,count
0,1970.0,1
1,2005.0,14
2,2006.0,44
3,2007.0,28
4,2008.0,29
5,2009.0,30
6,2010.0,19
7,2011.0,48
8,2012.0,41
9,2013.0,47


In [ ]:
# Count the number of channels created per year
channels_per_year = data['created_year'].value_counts().reset_index()
channels_per_year = channels_per_year.sort_values(by='created_year')

In [ ]:
fig_subscriber_growth = px.line(
    channels_per_year,
    x='created_year', y='count',
    title="Line Chart: Subscriber Growth Over the Years"
)
fig_subscriber_growth.show()

**Insights for Subscriber Growth Over the Years**
* **Consistent Growth:** Subscriber counts grew steadily from 1970 to 2000, reflecting a gradual adoption of content.
* **Rapid Expansion:** After 2000, there was a sharp increase in subscriber counts, peaking around 2010.
* **Recent Decline:** Subscriber growth has decreased after 2010, possibly due to market saturation or shifting content trends.
* **Historical Trends:** The chart highlights major milestones in subscriber adoption over decades, aligning with advancements in digital platforms.

**What is the yearly trend in the number of video uploads across all channels?**

In [ ]:
uploads_per_year = data.groupby('created_year')['uploads'].sum().reset_index()
uploads_per_year

,created_year,uploads
0,1970.0,744.0
1,2005.0,368526.0
2,2006.0,1350215.0
3,2007.0,1121837.0
4,2008.0,773802.0
5,2009.0,784175.0
6,2010.0,476397.0
7,2011.0,423894.0
8,2012.0,554507.0
9,2013.0,683046.0


In [ ]:
uploads_per_year = data.groupby('created_year')['uploads'].sum().reset_index()

# Plot the line chart
fig_uploads_trend = px.line(
    uploads_per_year,
    x='created_year', y='uploads',
    title="Line Chart: Total Video Uploads Over Years"
)
fig_uploads_trend.show()

**Insights for Line Chart: Subscriber Growth Over the Years**
* **Steady Increase:** Subscriber growth was steady from 1970 to 2000, indicating gradual adoption.
* **Peak Growth:** A sharp rise in subscriber numbers occurred post-2000, peaking around 2010.
* **Decline:** There has been a noticeable decline in growth after 2010, suggesting market saturation or changes in platform popularity.
* **Milestones:** Key growth spurts align with advancements in technology and platform evolution.

# ✨ Scatter Plot

## **What**
A scatter plot shows the relationship between two numerical variables.

## **Why**
- Identify correlations or trends between two variables.
- Visualize clusters or patterns in the data.
- Useful for regression analysis and anomaly detection.

## **Syntax**
```python
import plotly.express as px

fig = px.scatter(data_frame, x="numeric_column1", y="numeric_column2", color="category_column", title="✨ Scatter Plot")
fig.show()


**Is there a relationship between engagement rate and total video views across channels?**



In [ ]:
fig_engagement_vs_views = px.scatter(
    data, x='video views', y='engagement_rate',
    title="Scatter Plot: Engagement Rate vs Total Video Views",
    size='subscribers', color='Country',
    hover_name='Youtuber',
    labels={'video views': 'Total Video Views', 'engagement_rate': 'Engagement Rate'}
)
fig_engagement_vs_views.show()


**Insights for Scatter Plot: Engagement Rate vs. Total Video Views**
* **Inverse Relationship:** Countries with high total video views tend to have lower engagement rates.
* **Outliers:** Some countries achieve high engagement rates despite lower total views, indicating niche audience appeal.
* **Clusters:** The majority of countries fall within the lower engagement and view range, reflecting typical audience behavior.
* **Opportunity Areas:** Countries with mid-level views and engagement could be untapped for optimization.

**How does the number of subscribers correlate with video uploads?**

In [ ]:
fig_subscribers_vs_uploads = px.scatter(
    data, x='uploads', y='subscribers',
    title="Scatter Plot: Subscribers vs Video Uploads",
    size='engagement_rate', color='category',
    hover_name='Youtuber',
    labels={'uploads': 'Number of Uploads', 'subscribers': 'Subscribers'}
)
fig_subscribers_vs_uploads.show()


**Insights for Scatter Plot: Subscribers vs. Video Uploads**
* **High Impact Channels:** Some channels achieve high subscribers with relatively fewer uploads, showcasing quality over quantity.
* **Balanced Efforts:** Most channels with moderate subscribers have consistent upload numbers, reflecting steady content creation.
* **Outliers:** A few channels with extensive uploads but low subscribers may indicate inefficiency in content strategy.
* **Category Variation:** Different categories exhibit varying relationships between uploads and subscriber counts.

# 🥧 Pie Chart

## **What**
A pie chart represents proportions of categorical data as slices of a circle.

## **Why**
- Visualize part-to-whole relationships.
- Show proportions in survey results, demographics, or market shares.

## **Syntax**
```python
import plotly.express as px

fig = px.pie(data_frame, names="category_column", values="numeric_column", title="🥧 Pie Chart")
fig.show()


**What is the distribution of YouTube channels by category?**

In [ ]:
fig_category_pie = px.pie(
    data, names='category',
    title="Pie Chart: Distribution of YouTube Channels by Category",
    hole=0.6,  # Creates a donut chart
    labels={'category': 'Category'}
)
fig_category_pie.show()


**Insights for Pie Chart: Distribution of YouTube Channels by Category**
* **Entertainment Dominates:** Entertainment leads with 26.2%, showcasing its popularity.
* **Music and People & Blogs:** Music (19.9%) and People & Blogs (13%) also capture significant portions, reflecting audience interest in diverse topics.
* **Niche Categories:** Autos & Vehicles, Nonprofits, and Science & Technology are underrepresented, hinting at potential growth areas.
* **Balanced Distribution:** Categories like Gaming and Comedy have moderate shares, maintaining steady audience engagement.

**What is the Revenue Distribution Among Top 10 Highest-Earning Channels**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 554 entries, 0 to 553
Data columns (total 27 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Country                                  554 non-null    object 
 1   Gross tertiary education enrollment (%)  554 non-null    float64
 2   Population                               554 non-null    float64
 3   Title                                    554 non-null    object 
 4   Unemployment rate                        554 non-null    float64
 5   Urban_population                         554 non-null    float64
 6   Youtuber                                 554 non-null    object 
 7   category                                 554 non-null    object 
 8   channel_type                             554 non-null    object 
 9   channel_type_rank                        554 non-null    float64
 10  country_rank                             554 non-null  

In [ ]:
data.nlargest(10, 'highest_yearly_earnings')

,Country,Gross tertiary education enrollment (%),Population,Title,Unemployment rate,Urban_population,Youtuber,category,channel_type,channel_type_rank,...,lowest_yearly_earnings,rank,subscribers,subscribers_for_last_30_days,uploads,video views,video_views_for_the_last_30_days,video_views_rank,engagement_rate,views_per_upload
261,United States,88.2,3.282395e+08,DaFuq!?Boom!,14.70,270663028.0,DaFuq!?Boom!,Film & Animation,Entertainment,106.0,...,6900000.0,418.0,19600000.0,6700000.0,214.0,7.906182e+09,2.304000e+09,656.0,34.183673,3.694477e+07
0,India,28.1,1.366418e+09,T-Series,5.36,471031528.0,T-Series,Music,Music,1.0,...,6800000.0,1.0,245000000.0,2000000.0,20082.0,2.280000e+11,2.258000e+09,1.0,0.816327,1.135345e+07
2,United States,88.2,3.282395e+08,Cocomelon - Nursery Rhymes,14.70,270663028.0,Cocomelon - Nursery Rhymes,Education,Education,1.0,...,5900000.0,4.0,162000000.0,1000000.0,966.0,1.640000e+11,1.975000e+09,2.0,0.617284,1.697723e+08
3,India,28.1,1.366418e+09,SET India,5.36,471031528.0,SET India,Shows,Entertainment,2.0,...,5500000.0,5.0,159000000.0,1000000.0,116536.0,1.480000e+11,1.824000e+09,3.0,0.628931,1.269994e+06
13,India,28.1,1.366418e+09,Zee TV,5.36,471031528.0,Zee TV,Entertainment,Entertainment,8.0,...,5100000.0,22.0,70500000.0,900000.0,129204.0,7.313905e+10,1.707000e+09,9.0,1.276596,5.660742e+05
101,India,28.1,1.366418e+09,StarPlus,5.36,471031528.0,StarPlus,Entertainment,Entertainment,40.0,...,5000000.0,140.0,32000000.0,1200000.0,44892.0,2.680067e+10,1.667000e+09,60.0,3.750000,5.970034e+05
9,India,28.1,1.366418e+09,Sony SAB,5.36,471031528.0,Sony SAB,Shows,Entertainment,7.0,...,5000000.0,16.0,83000000.0,1100000.0,71270.0,1.010000e+11,1.657000e+09,4.0,1.325301,1.417146e+06
61,Brazil,51.3,2.125594e+08,GR6 EXPLODE,12.08,183241641.0,GR6 EXPLODE,Music,Music,30.0,...,4900000.0,84.0,38900000.0,100000.0,3043.0,2.515423e+10,1.635000e+09,73.0,0.257069,8.266261e+06
456,United States,88.2,3.282395e+08,ViralHog,14.70,270663028.0,ViralHog,People & Blogs,Animals,3.0,...,4100000.0,799.0,14000000.0,1100000.0,41117.0,1.891769e+10,1.364000e+09,136.0,7.857143,4.600941e+05
1,United States,88.2,3.282395e+08,MrBeast,14.70,270663028.0,MrBeast,Entertainment,Entertainment,1.0,...,4000000.0,3.0,166000000.0,8000000.0,741.0,2.836884e+10,1.348000e+09,48.0,4.819277,3.828454e+07


In [ ]:
top_earning_channels = data.nlargest(10, 'highest_yearly_earnings')
fig_highest_earnings_pie = px.pie(
    top_earning_channels, names='Youtuber', values='highest_yearly_earnings',
    title="Pie Chart: Revenue Distribution Among Top 10 Highest-Earning Channels",
    hole=0.4,  # Creates a donut chart
    labels={'highest_yearly_earnings': 'Yearly Earnings'}
)
fig_highest_earnings_pie.show()


**Insights for Pie Chart: Revenue Distribution Among Top 10 Highest-Earning Channels**
* **Diverse Revenue Spread:** Top channels like DaFuq!?Boom! (13%) and T-Series (12.7%) dominate revenue distribution, followed by others with smaller shares.
* **Balanced Earnings:** Revenue among top channels is relatively evenly spread, with no single channel overwhelmingly dominating.
* **Global Appeal:** Channels from different genres highlight the global diversity of high-earning content.
* **Opportunities:** Consistent revenue distribution across the top channels reflects opportunities for new entrants in lucrative content niches.







# 📊 Bar Plot

## **What**
A bar plot represents categorical data using rectangular bars where the length corresponds to the frequency or value.

## **Why**
- Compare values across different categories.
- Analyze trends or distributions in categorical data.
- Useful for visualizing group comparisons or survey results.

## **Syntax**
```python
import plotly.express as px

fig = px.bar(data_frame, x="category_column", y="numeric_column", title="📊 Bar Plot")
fig.show()


**Which categories have the highest total video views?**

In [ ]:
data.groupby('category')['video views'].sum().reset_index().sort_values(by='video views', ascending=False)

,category,video views
8,Music,1.937407e+12
3,Entertainment,1.813469e+12
11,People & Blogs,7.634447e+11
2,Education,6.107155e+11
14,Shows,4.379051e+11
4,Film & Animation,3.899961e+11
5,Gaming,3.893691e+11
1,Comedy,3.695646e+11
9,News & Politics,2.129565e+11
15,Sports,1.320809e+11


In [ ]:
category_views = data.groupby('category')['video views'].sum().reset_index().sort_values(by='video views', ascending=False)

fig_category_views_bar = px.bar(
    category_views, x='category', y='video views',
    title="Bar Chart: Total Video Views by Category",
    #text='video views',
    color='category'
)
fig_category_views_bar.show()


**Which countries have the highest total number of subscribers?**

In [ ]:
data.groupby('Country')['subscribers'].sum().reset_index().sort_values(by='subscribers', ascending=False)

,Country,subscribers
39,United States,4.546100e+09
14,India,3.828000e+09
3,Brazil,6.833000e+08
38,United Kingdom,5.420000e+08
15,Indonesia,3.838000e+08
26,Russia,3.575000e+08
30,South Korea,3.246000e+08
22,Mexico,3.093000e+08
0,Argentina,2.768000e+08
34,Thailand,2.723000e+08


In [ ]:
country_subscribers = data.groupby('Country')['subscribers'].sum().reset_index().sort_values(by='subscribers', ascending=False)

fig_country_subscribers_bar = px.bar(
    country_subscribers, x='Country', y='subscribers',
    title="Bar Chart: Total Subscribers by Country",
    #text='subscribers',
    color='Country'
)
fig_country_subscribers_bar.show()


# 🌳 Tree Map

## **What**
A tree map visualizes hierarchical data using nested rectangles. Each rectangle's size represents a numeric value, and the nesting reflects hierarchy.

## **Why**
- Display relationships between categories and subcategories.
- Visualize part-to-whole relationships.
- Useful for hierarchical data like sales by category and subcategory.

## **Syntax**
```python
import plotly.express as px

fig = px.treemap(data_frame, path=["category_column1", "category_column2"], values="numeric_column", title="🌳 Tree Map")
fig.show()


**How is the subscriber count distributed across categories and channel types?**

In [ ]:
data.head(5)

,Country,Gross tertiary education enrollment (%),Population,Title,Unemployment rate,Urban_population,Youtuber,category,channel_type,channel_type_rank,...,lowest_yearly_earnings,rank,subscribers,subscribers_for_last_30_days,uploads,video views,video_views_for_the_last_30_days,video_views_rank,engagement_rate,views_per_upload
0,India,28.1,1.366418e+09,T-Series,5.36,471031528.0,T-Series,Music,Music,1.0,...,6800000.0,1.0,245000000.0,2000000.0,20082.0,2.280000e+11,2.258000e+09,1.0,0.816327,1.135345e+07
1,United States,88.2,3.282395e+08,MrBeast,14.70,270663028.0,MrBeast,Entertainment,Entertainment,1.0,...,4000000.0,3.0,166000000.0,8000000.0,741.0,2.836884e+10,1.348000e+09,48.0,4.819277,3.828454e+07
2,United States,88.2,3.282395e+08,Cocomelon - Nursery Rhymes,14.70,270663028.0,Cocomelon - Nursery Rhymes,Education,Education,1.0,...,5900000.0,4.0,162000000.0,1000000.0,966.0,1.640000e+11,1.975000e+09,2.0,0.617284,1.697723e+08
3,India,28.1,1.366418e+09,SET India,5.36,471031528.0,SET India,Shows,Entertainment,2.0,...,5500000.0,5.0,159000000.0,1000000.0,116536.0,1.480000e+11,1.824000e+09,3.0,0.628931,1.269994e+06
4,Russia,81.9,1.443735e+08,Like Nastya Vlog,4.59,107683889.0,Like Nastya,People & Blogs,People,25.0,...,146800.0,9.0,106000000.0,100000.0,493.0,9.047906e+10,4.894700e+07,630.0,0.094340,1.835275e+08


In [ ]:
fig_treemap = px.treemap(data, path=['category', 'channel_type'],
                         values='subscribers',
                         title="Treemap: Subscriber Count by Category and Channel Type",
                         color='subscribers')
fig_treemap.show()


**Insights for Treemap: Subscriber Count by Category and Channel Type**
* **Entertainment Dominates:** Entertainment occupies the largest area, highlighting its overwhelming dominance in subscriber count.
* **Music as a Key Category:** Music follows closely, emphasizing its broad appeal and strong subscriber base.
* **Diverse Engagement:** Categories like People & Blogs, Gaming, and Education show moderate subscriber representation, indicating consistent audience interest.
* **Niche Potential:** Smaller sections like How-to & Style and News suggest potential for growth in these less-saturated areas.

# 🌡️ Heat Map

## **What**
A heat map visualizes relationships or correlations between variables using colors to represent values.

## **Why**
- Identify patterns, correlations, or clusters in data.
- Visualize the intensity of values across a matrix or grid.
- Useful for correlation matrices, geographic data, or density maps.

## **Syntax**
```python
import plotly.graph_objects as go

fig = go.Figure(data=go.Heatmap(
    z=data_frame.corr().values,
    x=data_frame.columns,
    y=data_frame.columns,
    colorscale="Viridis"  # Choose a colorscale like "Viridis", "Cividis", etc.
))
fig.update_layout(title="🌡️ Heat Map")
fig.show()


In [ ]:
data=go.Heatmap(
    z=data.corr().values,
    x=data.columns,
    y=data.columns,
    colorscale="Viridis"  # Choose a colorscale like "Viridis", "Cividis", etc.
)

data

ValueError: could not convert string to float: 'India'

In [ ]:
import plotly.graph_objects as go

fig = go.Figure(data=go.Heatmap(
    z=data.corr().values,
    x=data.columns,
    y=data.columns,
    colorscale="Viridis"  # Choose a colorscale like "Viridis", "Cividis", etc.
))
fig.update_layout(title="🌡️ Heat Map")
fig.show()

ValueError: could not convert string to float: 'India'

In [ ]:
data[['subscribers', 'video views', 'engagement_rate', 'uploads']].corr()

,subscribers,video views,engagement_rate,uploads
subscribers,1.000000,0.850776,-0.059782,0.074791
video views,0.850776,1.000000,-0.049922,0.147477
engagement_rate,-0.059782,-0.049922,1.000000,-0.021616
uploads,0.074791,0.147477,-0.021616,1.000000


In [ ]:
import plotly.figure_factory as ff

# Assuming 'data' is your dataframe
correlation = data[['subscribers', 'video views', 'engagement_rate', 'uploads']].corr()

fig_heatmap = ff.create_annotated_heatmap(
    z=correlation.values,
    y=list(correlation.columns),
    x=list(correlation.index),
    colorscale='Viridis'
)

# Adding title to the layout
fig_heatmap.update_layout(
    title="Heatmap: Correlation Between Numeric Variables",
    xaxis_title="Variables",
    yaxis_title="Variables"
)

fig_heatmap.show()


**Insights for Heatmap: Correlation Between Numeric Variables**
* **Strong Positive Correlation:** Subscribers and Video Views have a high correlation (0.85), indicating that channels with more subscribers generally have higher video views.
* **Weak Correlation with Uploads:** Uploads show a very weak correlation with Subscribers (0.07) and Video Views (0.15), suggesting that the quantity of uploads does not significantly drive these metrics.
* **Engagement Rate Independence:** Engagement Rate has near-zero or slightly negative correlations with other variables, showing that it operates independently of subscribers, views, and uploads.
* **Diagonal Values:** As expected, all variables show a perfect correlation (1.0) with themselves along the diagonal.

# 🌐 Bubble Chart

## **What**
A bubble chart adds a third dimension to a scatter plot using bubble sizes, representing an additional variable.

## **Why**
- Visualize relationships between three variables.
- Compare groups using size as an indicator.
- Useful for analyzing complex datasets, such as population vs. GDP vs. life expectancy.

## **Syntax**
```python
import plotly.express as px

fig = px.scatter(
    data_frame,
    x="numeric_column1",  # X-axis variable
    y="numeric_column2",  # Y-axis variable
    size="size_column",   # Bubble size
    color="category_column",  # Bubble color (categorical)
    hover_name="hover_column",  # Text displayed on hover
    title="🌐 Bubble Chart"
)
fig.show()


**How do video views, engagement rates, and subscriber counts compare across categories?**

In [ ]:
import plotly.express as px
fig_bubble = px.scatter(
    data, x='video views', y='engagement_rate',
    size='subscribers', color='category',
    title="Bubble Chart: Video Views vs Engagement Rate by Category",
    hover_name='Youtuber'
)
fig_bubble.show()


**Insights for Bubble Chart: Video Views vs. Engagement Rate by Category**
* **High Engagement, Low Views:** Certain categories achieve high engagement rates despite low total video views, indicating niche audience involvement.
* **Dominant View Categories:** Music and Entertainment dominate in video views but generally show lower engagement rates.
* **Clusters of Behavior:** Most categories cluster around low engagement and view counts, representing typical channel performance.
* **Outliers:** A few bubbles represent extreme outliers, showcasing categories or videos with exceptional popularity or engagement.

# **What is a Gauge Chart?**
A gauge chart is a circular visualization that represents a value as a percentage or fraction of a goal or range. It is widely used to show progress, performance, or key metrics against a defined target.

**Uses of a Gauge Chart:**

* Monitoring performance metrics, such as website traffic or sales targets.
* Tracking KPIs (Key Performance Indicators) like customer satisfaction or revenue goals.
* Indicating operational status, such as server performance or energy consumption.

## **show the engagement rate for a specific channel:**

In [ ]:
data

,Unnamed: 0,Country,Gross tertiary education enrollment (%),Population,Title,Unemployment rate,Urban_population,Youtuber,category,channel_type,...,lowest_yearly_earnings,rank,subscribers,subscribers_for_last_30_days,uploads,video views,video_views_for_the_last_30_days,video_views_rank,engagement_rate,views_per_upload
0,0.0,India,28.1,1.366418e+09,T-Series,5.36,471031528.0,T-Series,Music,Music,...,6800000.0,1.0,245000000.0,2000000.0,20082.0,2.280000e+11,2.258000e+09,1.0,0.816327,1.135345e+07
1,2.0,United States,88.2,3.282395e+08,MrBeast,14.70,270663028.0,MrBeast,Entertainment,Entertainment,...,4000000.0,3.0,166000000.0,8000000.0,741.0,2.836884e+10,1.348000e+09,48.0,4.819277,3.828454e+07
2,3.0,United States,88.2,3.282395e+08,Cocomelon - Nursery Rhymes,14.70,270663028.0,Cocomelon - Nursery Rhymes,Education,Education,...,5900000.0,4.0,162000000.0,1000000.0,966.0,1.640000e+11,1.975000e+09,2.0,0.617284,1.697723e+08
3,4.0,India,28.1,1.366418e+09,SET India,5.36,471031528.0,SET India,Shows,Entertainment,...,5500000.0,5.0,159000000.0,1000000.0,116536.0,1.480000e+11,1.824000e+09,3.0,0.628931,1.269994e+06
4,8.0,Russia,81.9,1.443735e+08,Like Nastya Vlog,4.59,107683889.0,Like Nastya,People & Blogs,People,...,146800.0,9.0,106000000.0,100000.0,493.0,9.047906e+10,4.894700e+07,630.0,0.094340,1.835275e+08
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
549,989.0,United States,88.2,3.282395e+08,Migos ATL,14.70,270663028.0,Migos ATL,Music,Entertainment,...,148200.0,990.0,12400000.0,100000.0,99.0,6.993406e+09,4.941200e+07,833.0,0.806452,7.064047e+07
550,990.0,Brazil,51.3,2.125594e+08,Natan por AÔø,12.08,183241641.0,Natan por AÔø,Sports,Entertainment,...,1700000.0,991.0,12300000.0,700000.0,1200.0,9.029610e+09,5.525130e+08,525.0,5.691057,7.524675e+06
551,991.0,India,28.1,1.366418e+09,Free Fire India Official,5.36,471031528.0,Free Fire India Official,People & Blogs,Games,...,194200.0,992.0,12300000.0,300000.0,1500.0,1.674410e+09,6.473500e+07,6141.0,2.439024,1.116273e+06
552,993.0,Sweden,67.0,1.028545e+07,RobTopGames,6.48,9021165.0,RobTopGames,Gaming,Games,...,11600.0,994.0,12300000.0,100000.0,39.0,3.741235e+08,3.871000e+06,35112.0,0.813008,9.592910e+06


In [ ]:
# Example: Gauge chart for engagement rate of a specific channel (e.g., T-Series)
import plotly.graph_objects as go
channel_name = "T-Series"
channel_data = data[data['Title'] == channel_name]
engagement_rate = channel_data['engagement_rate'].iloc[0]

fig_gauge = go.Figure(
    go.Indicator(
        mode="gauge+number",
        value=engagement_rate * 100,  # Convert to percentage
        gauge={
            "axis": {"range": [0, 100]},  # Min and Max in percentage
            "bar": {"color": "green"},
            "steps": [
                {"range": [0, 50], "color": "lightgray"},
                {"range": [50, 100], "color": "lightblue"}
            ]
        },
        title={"text": f"Engagement Rate for {channel_name}"}
    )
)

fig_gauge.show()


**Gauge Chart: Engagement Rate for T-Series**
* **High Engagement Rate:** T-Series has an impressive engagement rate of 81.6%, which demonstrates its ability to maintain strong interaction with its audience.
* **Industry Leadership:** The high engagement rate positions T-Series as a leader in audience engagement within its category.
* **Audience Loyalty:** The engagement rate indicates that the channel's content resonates well with its audience, fostering loyalty and repeated interaction.
* **Benchmark Surpassing:** If the industry standard engagement rate is around 70%, T-Series exceeds it by over 16%, showcasing exemplary performance.

# **What is a Waterfall Chart?**
A waterfall chart is a visualization tool used to display the cumulative effect of sequentially introduced positive or negative values. It is commonly used in financial and operational analysis to show how an initial value is influenced by a series of intermediate positive and negative changes, leading to a final value.

**Uses of a Waterfall Chart:**

* Understanding financial data, such as revenue breakdown or profit analysis.
* Visualizing changes in metrics over time or across categories.
*Explaining contributions to a total, such as product sales, operating expenses, etc.


General syntax


```
import plotly.graph_objects as go

fig = go.Figure(
    go.Waterfall(
        x=["Category A", "Category B", "Category C", "Total"],  # Categories
        y=[100, -20, 30, 110],  # Contributions (positive/negative)
        connector={"line": {"color": "gray"}}  # Style of connectors
    )
)

fig.update_layout(
    title="Waterfall Chart Example",
    xaxis_title="Categories",
    yaxis_title="Values"
)

fig.show()
```



# **show the video views breakdown by the top 5 channels**

In [ ]:
data["video views"].tolist()

[228000000000.0,
 28368841870.0,
 164000000000.0,
 148000000000.0,
 90479060027.0,
 77180169894.0,
 57856289381.0,
 77428473662.0,
 32144597566.0,
 101000000000.0,
 20826993957.0,
 30608119724.0,
 28634566938.0,
 73139054467.0,
 38843229963.0,
 45757850229.0,
 28648024439.0,
 61510906457.0,
 16241549158.0,
 59316472754.0,
 33431802698.0,
 57271630846.0,
 40602020243.0,
 25307753534.0,
 28837144516.0,
 27073872856.0,
 14696003229.0,
 35302243691.0,
 30367676736.0,
 30516172739.0,
 29884657286.0,
 24004842608.0,
 9877365274.0,
 14631710289.0,
 13626331061.0,
 30788679536.0,
 22519705183.0,
 27286058807.0,
 22936630813.0,
 10323391593.0,
 21388725229.0,
 16602198273.0,
 7828610828.0,
 41139050371.0,
 25458952022.0,
 50292540392.0,
 27568757295.0,
 36458726976.0,
 37939780685.0,
 26820902622.0,
 24519022988.0,
 22477745835.0,
 17608931161.0,
 5603111948.0,
 39450824833.0,
 23884824160.0,
 3294013141.0,
 10507474316.0,
 44900897958.0,
 16118181673.0,
 36131228583.0,
 25154232306.0,
 7339333

In [ ]:
import plotly.graph_objects as go

# Top 5 channels by video views
top_channels = data.nlargest(5, 'video views')
categories = top_channels['Title'].tolist() + ['Total']
video_views = top_channels['video views'].tolist()
video_views.append(sum(video_views))  # Add total

fig_waterfall = go.Figure(
    go.Waterfall(
        x=categories,
        y=video_views,
        connector={"line": {"color": "rgb(63, 63, 63)"}},
        text=[f"{int(val):,}" for val in video_views],
        textposition="outside"
    )
)

fig_waterfall.update_layout(
    title="Waterfall Chart: Video Views Breakdown for Top Channels",
    xaxis_title="Channels",
    yaxis_title="Video Views",
    showlegend=True
)

fig_waterfall.show()


**Waterfall Chart: Video Views Breakdown for Top Channels**
* **T-Series Dominance:** T-Series leads the chart with 228 billion video views, contributing the largest share to the total.
* **Cumulative Contribution:** The top five channels (T-Series, Cocomelon, SET India, Sony SAB, and Like Nastya Vlog) together account for over 731 billion views, illustrating the dominance of a few key players in the market.
* **Incremental Growth:** Each channel incrementally adds to the overall video views, highlighting the varying degrees of influence each has on the total.
* **Market Concentration:** A significant portion of video views is concentrated in the hands of a few top channels, indicating a market where major players dominate.



# What is a Subplot?
A **subplot** is a way to display multiple plots within a single figure, arranged in a grid structure. Each plot occupies a specific position in the grid, allowing for easy comparison of multiple visualizations.

In **Plotly**, subplots are created using the `make_subplots()` function, where you can define the number of rows, columns, and customization for each subplot.

---

### Keywords in the Code

#### **1. `make_subplots()`**
- **Purpose**: Creates a grid layout for multiple plots.
- **Arguments**:
  - `rows` and `cols`: Define the grid structure (e.g., `rows=2, cols=2` for a 2x2 grid).
  - `subplot_titles`: Adds titles above each subplot.

---

#### **2. Plot Types**
- **`go.Histogram`**: Creates a histogram for frequency distribution.
  - `x`: Data for the histogram (e.g., `data['subscribers']`).
  - `name`: The trace name (used in the legend).
  
- **`go.Scatter`**: Creates scatter or line plots.
  - `x`: Data for the x-axis (e.g., `data['rank']`).
  - `y`: Data for the y-axis (e.g., `data['video views']`).
  - `mode`: Specifies the style of the plot (`lines+markers` or `markers`).
  - `name`: The trace name (used in the legend).
  
- **`go.Bar`**: Creates a bar chart for comparison.
  - `x`: Categories for the bars (e.g., `engagement_data.index`).
  - `y`: Values for each bar (e.g., `engagement_data.values`).
  - `name`: The trace name (used in the legend).

---

#### **3. `fig.add_trace()`**
- **Purpose**: Adds a plot (trace) to a specific subplot.
- **Arguments**:
  - Plot type (`go.Histogram`, `go.Scatter`, etc.).
  - `row` and `col`: Position of the plot in the grid (e.g., `row=1, col=1` for the first subplot).

---

#### **4. `fig.update_xaxes()` and `fig.update_yaxes()`**
- **Purpose**: Customize the x-axis and y-axis for a specific subplot.
- **Arguments**:
  - `title_text`: Axis title (e.g., "Subscribers").
  - `row` and `col`: Subplot location to apply changes.

---

#### **5. `fig.update_layout()`**
- **Purpose**: Configures the overall layout of the figure.
- **Arguments**:
  - `height` and `width`: Dimensions of the figure.
  - `title_text`: Main title for the entire figure.
  - `showlegend`: Shows or hides the legend.

---

#### **6. `groupby()`**
- **Purpose**: Groups data by a specific column and applies aggregation (like mean or sum).
- **Example**:
  ```python
  engagement_data = data.groupby('category')['engagement_rate'].mean()


In [ ]:
# Import necessary modules from Plotly
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Create a subplot layout with 2 rows and 2 columns
# Add custom titles for each subplot
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=("Histogram of Subscribers",  # Top-left: Histogram
                    "Line Plot: Video Views vs Rank",  # Top-right: Line Plot
                    "Bar Chart: Engagement Rate by Category",  # Bottom-left: Bar Chart
                    "Scatter Plot: Subscribers vs Video Views")  # Bottom-right: Scatter Plot
)

# Add a Histogram to the first subplot (row 1, col 1)
fig.add_trace(
    go.Histogram(x=data['subscribers'], name="Subscribers Histogram"),  # Plot subscribers as a histogram
    row=1, col=1  # Specify the location in the subplot grid
)
# Set x-axis and y-axis labels for the histogram
fig.update_xaxes(title_text="Subscribers", row=1, col=1)
fig.update_yaxes(title_text="Frequency", row=1, col=1)

# Add a Line Plot to the second subplot (row 1, col 2)
fig.add_trace(
    go.Scatter(x=data['rank'], y=data['video views'], mode='lines+markers', name="Video Views vs Rank"),  # Line plot with markers
    row=1, col=2  # Specify the location in the subplot grid
)
# Set x-axis and y-axis labels for the line plot
fig.update_xaxes(title_text="Rank", row=1, col=2)
fig.update_yaxes(title_text="Video Views", row=1, col=2)

# Prepare data for the Bar Chart: Calculate average engagement rate by category
engagement_data = data.groupby('category')['engagement_rate'].mean().sort_values(ascending=False)

# Add a Bar Chart to the third subplot (row 2, col 1)
fig.add_trace(
    go.Bar(x=engagement_data.index, y=engagement_data.values, name="Engagement Rate by Category"),  # Bar chart with categories and average engagement rates
    row=2, col=1  # Specify the location in the subplot grid
)
# Set x-axis and y-axis labels for the bar chart
fig.update_xaxes(title_text="Category", row=2, col=1)
fig.update_yaxes(title_text="Engagement Rate", row=2, col=1)

# Add a Scatter Plot to the fourth subplot (row 2, col 2)
fig.add_trace(
    go.Scatter(x=data['subscribers'], y=data['video views'], mode='markers',  # Scatter plot with subscribers and video views
               name="Subscribers vs Video Views"),
    row=2, col=2  # Specify the location in the subplot grid
)
# Set x-axis and y-axis labels for the scatter plot
fig.update_xaxes(title_text="Subscribers", row=2, col=2)
fig.update_yaxes(title_text="Video Views", row=2, col=2)

# Update the overall layout of the figure
fig.update_layout(
    height=800,  # Set the height of the entire figure
    width=1000,  # Set the width of the entire figure
    title_text="Subplot with 4 Different Graphs",  # Title of the figure
    showlegend=True  # Enable the legend to show plot labels
)

# The plot can now be displayed using fig.show() or saved as an HTML file


# **What is Interactivity?**
Interactivity in data visualization refers to the ability to interact dynamically with a chart or graph to explore, filter, or drill down into data. Features like dropdown menus, zooming, panning, and tooltips are examples of interactivity. Plotly provides interactive elements that let users:

* Filter Data: Use dropdowns or sliders to focus on specific subsets of data.
* Hover: See detailed information by hovering over data points.
* Zoom/Pan: Explore specific regions of a plot dynamically.

In [ ]:
data.groupby(['category', 'Country'])['video views'].sum().reset_index()

,category,Country,video views
0,Autos & Vehicles,Brazil,9.597895e+09
1,Autos & Vehicles,United Arab Emirates,5.405563e+09
2,Comedy,Argentina,2.813319e+10
3,Comedy,Brazil,3.052858e+10
4,Comedy,India,8.598253e+10
...,...,...,...
136,Sports,Brazil,9.029610e+09
137,Sports,Spain,5.166281e+09
138,Sports,Switzerland,5.529132e+09
139,Sports,United States,1.027548e+11


In [ ]:


# Aggregate data by category and country to sum video views
category_country_views = data.groupby(['category', 'Country'])['video views'].sum().reset_index()

# Create the initial bar chart
fig = px.bar(
    category_country_views,
    x='category',
    y='video views',
    color='category',
    title="Total Video Views by Category",
    labels={'video views': 'Video Views', 'category': 'Category'}
)

# Add dropdown for countries (multi-select logic implemented manually)
unique_countries = category_country_views['Country'].unique()

# Create a dropdown for country selection
fig.update_layout(
    updatemenus=[
        dict(
            buttons=[
                dict(label="All Countries",
                     method="update",
                     args=[{"visible": [True] * len(category_country_views)},
                           {"title": "Total Video Views by Category (All Countries)"}]),
                *[
                    dict(label=country,
                         method="update",
                         args=[{"visible": category_country_views['Country'] == country},
                               {"title": f"Total Video Views by Category ({country})"}])
                    for country in unique_countries
                ],
                dict(label="Clear",
                     method="update",
                     args=[{"visible": [False] * len(category_country_views)},
                           {"title": "Total Video Views by Category (No Country Selected)"}])
            ],
            direction="down",
            showactive=True,
            x=0.1,
            y=1.15,
        )
    ]
)

# Show the chart
fig.show()


# 📊 **Overall Summary of the Code**

This code creates an **interactive bar chart** using Plotly, allowing users to explore total video views by category with the ability to filter the chart by country using a dropdown menu. Here's the overall functionality:

---

## ✨ **Key Objectives of the Code**

### 🔍 **Summarization of Data**
- The data is grouped by **category** and **Country**, and the total `video views` are aggregated for each combination.  
- ✅ This ensures the visualization is **concise** and focused on **relevant metrics**.

---

### 📊 **Dynamic Visualization**
- A **bar chart** is created with:
  - **Categories** on the x-axis.
  - **Total video views** on the y-axis.
- 🎨 Each category is **color-coded** for better differentiation.

---

### 🎛️ **Interactive Features**
- A **dropdown menu** is added to allow filtering the data by **Country**.
- Users can toggle between viewing:
  - 🌍 **All Countries**
  - 🗺️ A **specific country**
  - ❌ A **clear option** to hide all data.

---

### 🔄 **Dynamic Updates**
- When a country is selected, the chart dynamically updates:
  - 👁️ Only the data for the **selected country** is displayed.
  - 📝 The chart **title** changes to reflect the selected filter.

---

### 👨‍💻 **User-Friendly Interface**
- 📐 Dropdown positioning and options are designed for **easy navigation**.  
- 🏷️ Labels, titles, and tooltips are added to improve **readability**.

---

## ✅ **Conclusion**
This code effectively demonstrates Plotly's ability to build **interactive and engaging visualizations**. It provides a solid foundation for exploring data dynamically and can be easily adapted for other use cases with filters and metrics. 🚀


### 🔍 Activity: True or False

**Q1:** Plotly is a Python library exclusively used for creating static visualizations.  
- **Answer:** **False**  
  **Explanation:** Plotly is primarily used for creating **interactive visualizations**, not just static ones.

---

**Q2:** Plotly supports integration with pandas, allowing direct plotting from DataFrame columns.  
- **Answer:** **True**  
  **Explanation:** Plotly integrates seamlessly with pandas, making it easy to create plots using column names.

---

**Q3:** The `plotly.express` module is used for creating high-level, simplified visualizations.  
- **Answer:** **True**  
  **Explanation:** The `plotly.express` module provides a concise and high-level interface for creating common visualizations.

---

**Q4:** The `update_layout()` method in Plotly cannot be used to customize the chart title.  
- **Answer:** **False**  
  **Explanation:** The `update_layout()` method allows customization of titles, axes, legends, and more.

---

**Q5:** Plotly's `go.Figure` object is used for low-level, highly customizable visualizations.  
- **Answer:** **True**  
  **Explanation:** The `go.Figure` object in the `plotly.graph_objects` module provides full control for creating and customizing plots.


# 📊 **Summary: Plotly**

## 🌟 **What Did We Learn?**

1. **Interactive Visualizations**:
   - Plotly excels in creating dynamic, interactive charts, including bar charts, scatter plots, line graphs, and more.
   - It allows users to explore data interactively with zoom, hover, and filtering options.

2. **Integration with Pandas**:
   - Plotly works seamlessly with pandas DataFrames, enabling easy plotting directly from data columns.

3. **High-Level vs Low-Level APIs**:
   - **`plotly.express`**: High-level interface for quick and simple visualizations.
   - **`plotly.graph_objects`**: Low-level interface for detailed customization and control over chart elements.

4. **Customizations**:
   - Advanced customization options are available using methods like `update_layout()`, `update_traces()`, and others.
   - You can modify titles, axes, colors, legends, and even interactivity settings.

5. **Wide Range of Chart Types**:
   - Plotly supports 2D and 3D plots, maps, animations, and statistical visualizations, making it highly versatile for diverse data.

6. **Dash Integration**:
   - Plotly works with Dash for building web-based, interactive dashboards seamlessly.

---

## ⚠️ **Shortcomings and Limitations**

1. **Learning Curve**:
   - Advanced customizations and working with `graph_objects` can be complex for beginners.

2. **Static Output**:
   - By default, Plotly outputs interactive visualizations. Static visualizations may require additional configurations or exports.

3. **Performance**:
   - Rendering issues may arise with large datasets due to browser limitations, affecting responsiveness.

4. **Dependency on JavaScript**:
   - Plotly's interactivity depends on JavaScript, which can be limiting in environments without proper support.

5. **Limited Multi-Select in Native Dropdowns**:
   - Plotly’s native dropdowns don’t support multi-select options directly, requiring Dash or workarounds.





# 📈 **Best Practices and Tips for Plotly**

## 🌟 **Best Practices**
1. **Start with `plotly.express`:**
   - Use `plotly.express` for quick visualizations and default aesthetics. Switch to `plotly.graph_objects` for advanced customization.

2. **Aggregate Data:**
   - Pre-process and aggregate data (e.g., using `pandas.groupby`) to avoid cluttered charts and improve performance.

3. **Use Efficient Layouts:**
   - Optimize the layout using `update_layout()` to improve readability (titles, axes labels, and legends).

4. **Limit Data Size:**
   - For large datasets, sample or aggregate the data to avoid performance issues in rendering interactive plots.

5. **Keep Charts Accessible:**
   - Add appropriate labels, tooltips, and titles to make charts user-friendly and understandable.

6. **Integrate Dash for Complex Interactions:**
   - Use Dash to build interactive dashboards with filters, dropdowns, and callbacks for advanced user interactions.

7. **Choose the Right Chart Type:**
   - Select chart types that best represent the data (e.g., bar for categories, scatter for relationships, heatmap for correlations).

---

## 💡 **Enhance Your Knowledge Mnemonic**
**"PLOTLY"**  
- **P**: **Prepare** your data (clean, aggregate, and format it).  
- **L**: **Learn** both `plotly.express` for simplicity and `graph_objects` for advanced features.  
- **O**: **Optimize** your plots with efficient layouts and customizations.  
- **T**: **Test** interactivity (zoom, hover, and filters) to ensure a seamless experience.  
- **L**: **Limit** data size to improve rendering performance.  
- **Y**: **Your Story** matters—design plots to effectively communicate insights.

---

## 🔗 **References and Resources**
1. **Official Documentation**:
   - [Plotly Python Library](https://plotly.com/python/)
   - [Plotly Express Documentation](https://plotly.com/python/plotly-express/)

2. **Dash (Interactive Dashboards)**:
   - [Dash by Plotly](https://dash.plotly.com/)

3. **Community and Tutorials**:
   - [Plotly Community Forum](https://community.plotly.com/)
   - [Plotly Tutorials on Medium](https://medium.com/tag/plotly)

4. **Advanced Plotly Customizations**:
   - [Customizing Layouts](https://plotly.com/python/figure-labels/)
   - [Interactive Dropdowns](https://plotly.com/python/dropdowns/)

5. **YouTube Tutorials**:
   - [Corey Schafer: Interactive Data Visualizations with Plotly](https://www.youtube.com/watch?v=Xs2ZT9wJ6hc)
   - [StatQuest: Intro to Plotly](https://www.youtube.com/results?search_query=statquest+plotly)

6. **Books**:
   - *Interactive Data Visualization with Python* by Abha Belorkar and Sharath Kumar M

---

With these practices and resources, you'll master Plotly and create stunning visualizations to effectively communicate insights! 🚀


---

## 💡 **Conclusion**

Plotly is a powerful library for creating interactive and visually appealing charts. Its integration with pandas and support for a variety of chart types make it a go-to tool for exploratory data analysis and dashboards. However, mastering advanced customizations and addressing performance limitations requires practice and careful planning.

# **TASKS**

**Sample data**

In [ ]:
import pandas as pd

# Sample dataset: YouTube Channel Metrics
data = {
    "Channel": ["TechTalk", "FoodiesHub", "MusicMania", "GameZone", "TravelVibes"],
    "Category": ["Technology", "Food", "Music", "Gaming", "Travel"],
    "Subscribers": [150000, 250000, 350000, 450000, 200000],
    "EngagementRate": [0.07, 0.12, 0.09, 0.11, 0.08],
    "VideoViews": [15000000, 25000000, 50000000, 30000000, 20000000],
    "Revenue": [1200, 2300, 4500, 3100, 1900]
}

df = pd.DataFrame(data)
df


,Channel,Category,Subscribers,EngagementRate,VideoViews,Revenue
0,TechTalk,Technology,150000,0.07,15000000,1200
1,FoodiesHub,Food,250000,0.12,25000000,2300
2,MusicMania,Music,350000,0.09,50000000,4500
3,GameZone,Gaming,450000,0.11,30000000,3100
4,TravelVibes,Travel,200000,0.08,20000000,1900


# **Tasks for Plotly Visualizations**
##**Task 1:** **Analyze Subscriber and Engagement Trends**
1. Bar Chart: Create a bar chart to compare subscriber counts across different categories.
2. Scatter Plot: Visualize the relationship between engagement rate and video views. Use bubble sizes to represent subscriber counts.

## **Task 2: Revenue Insights**
1. Pie Chart: Create a pie chart to show the revenue contribution of each channel.
2. Histogram: Visualize the frequency distribution of revenue using a histogram.

# **Task 3: Engagement and Revenue Comparison**
1. Box Plot: Compare engagement rates across different categories using a box plot.
2. Line Chart: Show the trend of revenue vs. subscribers.